### PR TITLE
feat: build cache

### LINKED ISSUE #6: cache facility for the build directory
Allow projects to configure directories to keep in cache after builds / restore from cache before build based on the contents of a file (eg. node_modules based on yarn.lock)

### PR DESCRIPTION
**Pull request by njoyard**
Add the ability to store files in a cache after a build, and restore them in future builds if the sha256 of a given file is the same. Cache has an expiration time so that old cached things don't stay around forever.

Use case: store node_modules based on the contents of yarn.lock

Also includes a small reorganization of the data directory.

Closes #6

### CODE REVIEW

File: config.sample.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
     "keyPassword": ""
   },
 
+  "cacheValidity": 2592000000,
```
Comment (xcambar) [inline config.sample.json]: Please find a more explicit name. `cache` could be many many things.

Comment (njoyard) [reply config.sample.json]: This is a difficult one :smile: 

buildCache ? assetCache  ? buildAssetCache ?

They seem equally generic to me....

**Comment by njoyard** [comment]
> why not deliver it in a more standalone fashion

I'm not sure I understand... You mean a third party package? The feature set seems way to thin for this IMO. We're just computing a sha256 and storing stuff as tgz here. The equivalent in bash would be 4 lines. Not worth its own package IMO :wink:

**Comment by xcambar** [comment]
Maybe I was too quick to read, but the `lib/cache.js` file is still worth a share, IMHO.

**Comment by njoyard** [comment]
Okay, why not then :slightly_smiling_face: 

You prefer I'd start a project with that and change this PR? Or we merge this and extract that feature later on?

**Comment by xcambar** [comment]
Let's merge and extract after

**Pull request merged**